lib/ansible/modules/cloud/cloudstack/cs_storage_pool.py

```diff
@@ -0,0 +1,399 @@
+
+        self.result['changed'] = True
+
+        id = None
+        if not self.module.check_mode:
+            res = self.cs.createStoragePool(**args)
+            if 'errortext' in res:
+                self.module.fail_json(msg="Failed: '%s'" % res['errortext'])
+            pool = res['storagepool']
+        return pool
+
+    def _update_storage_pool(self):
+        pool = self.get_storage_pool()
+
+        args = {}
+
+        args['id'] = pool['id']
+        args['capacitybytes'] = self.module.params.get('capacity_bytes')
+        args['capacityiops'] = self.module.params.get('capacity_iops')
+        args['tags'] = self.module.params.get('tags')
+        state = self.module.params.get('state')
+        if state in ['enabled', 'disabled']:
+            args['state'] = state.capitalize()
+
+        if self.has_changed(args, pool):
+            self.result['changed'] = True
+
+            if not self.module.check_mode:
+                res = self.cs.updateStoragePool(**args)
+                if 'errortext' in res:
+                    self.module.fail_json(msg="Failed: '%s'" % res['errortext'])
+        return pool
+
+    def absent_storage_pool(self):
+        pool = self.get_storage_pool()
+        if pool:
+            self.result['changed'] = True
+
+            args = {
+                'id': pool['id'],
+            }
+            if not self.module.check_mode:
+                res = self.cs.deleteStoragePool(**args)
+                if 'errortext' in res:
+                    self.module.fail_json(msg="Failed: '%s'" % res['errortext'])
+        return pool
+
+    def get_pod(self, key=None):
+        pod = self.module.params.get('pod')
+        if not pod:
+            return None
+        args = {
+            'name': pod,
+            'zoneid': self.get_zone(key='id'),
+        }
+        pods = self.cs.listPods(**args)
+        if pods:
+            return self._get_by_key(key, pods['pod'][0])
+        self.module.fail_json(msg="Pod %s not found in zone %s." % (
+            self.module.params.get('pod'),
+            self.get_zone(key='name')))
+
+    def get_cluster(self, key=None):
+        cluster_name = self.module.params.get('cluster')
+        if not cluster_name:
+            return None
+        args = {
+            'name': cluster_name,
+            'zoneid': self.get_zone(key='id'),
+        }
+        clusters = self.cs.listClusters(**args)
+        if clusters:
+            return self._get_by_key(key, clusters['cluster'][0])
+        self.module.fail_json(msg="Cluster %s not found" % cluster_name)
+
+
+def main():
+    argument_spec = cs_argument_spec()
+    argument_spec.update(dict(
+        name=dict(required=True),
+        storage_url=dict(default=None),
+        zone=dict(required=True),
+        pod=dict(default=None),
+        cluster=dict(default=None),
+        scope=dict(default=None, choices=['ZONE', 'CLUSTER']),
+        hypervisor=dict(default=None),
+        provider=dict(default='DefaultPrimary'),
```

Comment:
I don't think we can. The API docs don't give a list, and I suspect that this might be a pluggable system whereby any third party could write their own provider. Having said that I've done no research into this.